### DISCUSSION TITLE
Add tool to keep addon's public template-registry in sync

### DISCUSSION THREAD

**Discussion by phndiaye** [typed-ember/glint#664] [Ideas]
In Glint's "Authoring addons" guide, it is said that when using classic `.hbs` files, we should expose a `addon/template-registry.d.ts` file that exports the addon's registry so  apps can extend it. That file is supposed to list all components/helpers/modifiers that the addon exposes to its parents.

However, when dealing with addons with a lot of those, things can easily get out of sync and you can end up forgetting to expose a component's Glint types.

That's why i suggest to have an additional script (`sync-template-registry`?) that goes through a project and automatically generates the `template-registry` file for us. It would:

- check for existing `template-registry.ts` file 
- go through the components/helpers/modifiers in a directory (via a glob pattern)
- build a diff so we know if a sync is needed
  - add missing declarations to the `template-registry.ts`
  - remove those that do not exist anymore (this may be optional as technically TS won't let an invalid import slide anyway)

```console
foo@bar:~$ npx -p @glint/scripts sync-template-registry addon/components/**/*.ts --help

Positionals:
  globs  One or more paths or globs specifying the files to act on.                         [string]

Options:
  --help              Show help [boolean]
  --version           Show version number [boolean]
  --registry-file     The path to the template registry file [string] [default: "addon/template-registry.ts"]
  --check             Returns whether or not the registry is out of sync [boolean]
```

Note that the default value for the `registry-file` option is based on a v1 addon as for v2 addons we can't really know how their author will decide to structure them.

The `--check` option can come in handy on a CI to prevent from publishing an addon w/ an out-of-sync template-registry.

**Comment by wagenet** [comment]
I've done something similar to this in AuditBoard's own app. It's not super polished, but I'll see if I can clean it up or at least share the basic approach I took.

**Reply by phndiaye** [reply]
@wagenet thanks! (and glad to know there could be some interest for it)

**Reply by phndiaye** [reply]
Hello @wagenet , thought about this more and [drafted something](https://github.com/typed-ember/glint/compare/main...phndiaye:glint:scripts/sync-template-registry?expand=1). It needs some polishing but it helped me generate everything for one of my projects already. Wondering if it resembles what you've done in AuditBoard's own app :)

**Comment by phndiaye** [comment]
@bertdeblock released https://github.com/bertdeblock/generate-template-registry which does the trick so I think i'll close this discussion :) 

Thanks again @bertdeblock! 💯 🥳
